### DISCUSSION TITLE
How to use number field modifier in Go?

### DISCUSSION THREAD

**Discussion by Oggy4th** [pocketbase/pocketbase#1974] [Q&A] [answered]
I am using pocketbase as framework and I wrote test API request with **NUMBER field modifier** to decrement wallet balance one by one.

I called this request 2000 times from **jmeter**. Wallet total balance is 2000 and after testing it suppose to be 0. When I sending requests with this config it works fine.
![image](https://user-images.githubusercontent.com/44497354/222951730-486c4aff-f8ac-4465-b5d5-1ee952fa3330.png)
![image](https://user-images.githubusercontent.com/44497354/222951649-84758eb4-59fd-435d-b766-3272be4ffa43.png)

Then I changed config and send it again, it doesn't work as I imagined.
![image](https://user-images.githubusercontent.com/44497354/222951899-7e473a72-7828-4136-85cb-911cf109f9c2.png)
![image](https://user-images.githubusercontent.com/44497354/222951918-faf57ff2-f71a-4cf3-bfb1-c286e50ccbf9.png)

It gave 2000 responses but didn't decreased most of the value like this one. 

![image](https://user-images.githubusercontent.com/44497354/222951943-10b9fc1a-e1a9-41c3-89f1-a6a443e6f986.png)

Below golang code is my test request.
```
func test(c echo.Context) error {
	record, err := database.App.Dao().FindRecordById("wallets", "RECORD_ID")
	if err != nil {
		return c.JSON(http.StatusBadRequest, err.Error())
	}

	form := forms.NewRecordUpsert(database.App, record)

	form.LoadData(map[string]any{
		"balance-": 1,
	})

	if err := form.Submit(); err != nil {
		return c.JSON(http.StatusBadRequest, err.Error())
	}

	return c.JSON(http.StatusOK, "ok")
}
```

I don't know where am I making mistakes. Is it right or wrong to use `form.LoadData` to decrease wallet balance?

**Comment by ganigeorgiev** [comment, marked as answer]
This is expected if the requests are executed concurrently. `"balance-": 1` is equivalent to `record.Set("balance", record.GetInt("balance") - 1)` (_btw. if you don't need form validations, you can [directly use the `Dao` to update your record](https://pocketbase.io/docs/record-methods/#update-record-without-data-validations)_).

I'm not sure that I understand your use case, but if you want to execute concurrent requests and to keep the number field in "sync" you'll have to implement a global mutex OR wrap the `test()` function content inside a [transaction](https://pocketbase.io/docs/record-methods/#transaction) (_instead of `database.App.Dao()` you'll have to use the `txDao` from the `RunInTransaction` callback_), but again I'm not sure that I understand what is the end goal here so the above may not be suitable for your use case.

**Reply by ghost** [reply]
I've been having same issue, I thought field modifiers were atomic operation. Does transaction have performance impact?

**Reply by Oggy4th** [reply]
Hello there @ganigeorgiev 

Thanks for your advice

Like you said I used Transaction. It's working now.

**Reply by ganigeorgiev** [reply]
> @temzints  Does transaction have performance impact?

Transactions (especially the mixed read/write) perform a db lock preventing multiple db operations to run concurrently, so Yes, it could impact the performance and this is why they are kept at minimum in PocketBase. Use them if you really need the transactional safety, otherwise you may be better without them.

**Reply by ghost** [reply]
Thank you very much for the explanation and I really appreciate this great open source project. It's just so convenient than other alternatives so much possibilities.
